xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go

```diff
@@ -852,6 +856,225 @@ func (s) TestEDS_ClusterResourceUpdates(t *testing.T) {
+func (s) TestEDS_BadUpdateWithPreviousGoodUpdate(t *testing.T) {
+	// Spin up a management server to receive xDS resources from.
+	mgmtServer, nodeID, bootstrapContents, _, cleanup1 := e2e.SetupManagementServer(t, e2e.ManagementServerOptions{})
+	defer cleanup1()
+
+	// Start a backend server that implementats the TestService.
+	server := stubserver.StartTestService(t, nil)
+	defer server.Stop()
+
+	// Create an EDS resource for consumption by the test.
+	resources := clientEndpointsResource(nodeID, edsServiceName, []e2e.LocalityOptions{{
+		Name:     localityName1,
+		Weight:   1,
+		Backends: []e2e.BackendOptions{{Port: testutils.ParsePort(t, server.Address)}},
+	}})
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Create an xDS xdsClient for use by the cluster_resolver LB policy.
+	xdsClient, close, err := xdsclient.NewWithBootstrapContentsForTesting(bootstrapContents)
+	if err != nil {
+		t.Fatalf("Failed to create xDS client: %v", err)
+	}
+	defer close()
+
+	// Create a manual resolver and push a service config specifying the use of
+	// the cluster_resolver LB policy with a single discovery mechanism.
+	r := manual.NewBuilderWithScheme("whatever")
+	jsonSC := fmt.Sprintf(`{
+			"loadBalancingConfig":[{
+				"cluster_resolver_experimental":{
+					"discoveryMechanisms": [{
+						"cluster": "%s",
+						"type": "EDS",
+						"edsServiceName": "%s",
+						"outlierDetection": {}
+					}],
+					"xdsLbPolicy":[{"round_robin":{}}]
+				}
+			}]
+		}`, clusterName, edsServiceName)
+	scpr := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(jsonSC)
+	r.InitialState(xdsclient.SetClient(resolver.State{ServiceConfig: scpr}, xdsClient))
+
+	// Create a ClientConn and make a successful RPC.
+	cc, err := grpc.Dial(r.Scheme()+":///test.service", grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithResolvers(r))
+	if err != nil {
+		t.Fatalf("failed to dial local test server: %v", err)
+	}
+	defer cc.Close()
+
+	// Ensure RPCs are being roundrobined across the single backend.
+	client := testgrpc.NewTestServiceClient(cc)
+	if err := rrutil.CheckRoundRobinRPCs(ctx, client, []resolver.Address{{Addr: server.Address}}); err != nil {
+		t.Fatal(err)
+	}
+
+	// Update the endpoints resource in the management server with a load
+	// balancing weight of 0. This will result in the resource being NACKed by
+	// the xDS client. But since the cluster_resolver LB policy has a previously
+	// received good EDS update, it should continue using it.
+	resources.Endpoints[0].Endpoints[0].LbEndpoints[0].LoadBalancingWeight = &wrapperspb.UInt32Value{Value: 0}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Ensure that RPCs continue to succeed for the next second.
+	for end := time.Now().Add(time.Second); time.Now().Before(end); <-time.After(defaultTestShortTimeout) {
```

Comment:
Same thing here wrt 1 second additional polling time.

Reply:
Again, this being an e2e test, there is no way for us to ensure that the update we pushed to the management server was received by the xDS client and the configuration plumbed all the way down to all the LB policies, and the picker updated etc. So, we have to rely on some heuristic. We have been using this model a lot in these e2e tests, and yes, it does add a second of delay to the test, but it does ensure that the test is not flaky (because it was initially before I resorted to this model).

Reply:
Alright, sgtm.